### ISSUE TITLE
Techdocs can't query catalog metadata when using app-backend

### ISSUE THREAD

**Issue by taras** [backstage/backstage#3279]
<img width="1638" alt="image" src="https://user-images.githubusercontent.com/74687/98888201-86651380-2465-11eb-860f-bd37f17575c6.png">

When using app-backend where frontend and backend are running off the same URL, Techdocs is not able to retrieve metadata and fails.

## Expected Behavior

Should be able to use Techdocs in app-backend.

## Current Behavior

For some reason, fetch to `${catalogUrl}/entities/by-name/${kind}/${namespace}/${name}` returns content of `index.html` instead of metadata. I can use the generated URL to retrieve the metadata by calling the API directly via the browser.

## Possible Solution

🤷‍♀️ 

## Steps to Reproduce

1. Using release 0.2.0

```yaml
app:
  baseUrl: http://localhost:7000

backend:
  baseUrl: http://localhost:7000
  listen:
    port: 7000
  csp:
    connect-src: ["'self'", 'http:', 'https:']

techdocs:
  storageUrl: http://localhost:7000/api/techdocs/static/docs
  requestUrl: http://localhost:7000/api/techdocs
```

2. Use the config above on the example app and backend
2. Try to generate documentation
3. You should get a similar error as above

## Context

Trying to get a basic documentation site working for demo purposes. 

## Your Environment

- NodeJS Version (v12): 12
- Operating System and Version (e.g. Ubuntu 14.04): Mac OS X
- Browser Information: Chrome

**Comment by hooloovooo** [comment]
Have you tried this in another browser like Firefox? We've had some issues that are because of chrome caching some stuff really aggressively and just want to make sure we can exclude that.

**Comment by Rugvip** [comment]
@taras Do you have the backend routes set up like the example backend? https://github.com/backstage/backstage/blob/d5a2bf9296c6a299d59df4f05023a8313143a375/packages/backend/src/index.ts#L101

It's important to have the separate `/api` router, including the `notFoundHandler()` as the last middleware of the `apiRouter`.

**Comment by taras** [comment]
@Rugvip I'm testing this inside of Backstage repository on v0.2.0 tag. Here is what Router setup looks like,

```ts
const apiRouter = Router();
  apiRouter.use('/catalog', await catalog(catalogEnv));
  apiRouter.use('/rollbar', await rollbar(rollbarEnv));
  apiRouter.use('/scaffolder', await scaffolder(scaffolderEnv));
  apiRouter.use('/sentry', await sentry(sentryEnv));
  apiRouter.use('/auth', await auth(authEnv));
  apiRouter.use('/techdocs', await techdocs(techdocsEnv));
  apiRouter.use('/kubernetes', await kubernetes(kubernetesEnv));
  apiRouter.use('/proxy', await proxy(proxyEnv));
  apiRouter.use('/graphql', await graphql(graphqlEnv));
  apiRouter.use(notFoundHandler());

  const service = createServiceBuilder(module)
    .loadConfig(config)
    .addRouter('', await healthcheck(healthcheckEnv))
    .addRouter('/api', apiRouter)
    .addRouter('', await app(appEnv));
```

I just ran `yarn clean`, `yarn tsc` & `yarn build` and checked Safari, Chrome & Firefox.

**Comment by jennisays** [comment]
The router is missing an entry which could be why it's serving index.html. So... isn't an easy/obvious fix to add a route that matches instead of the fallback to index.html?

**Comment by taras** [comment]
@jennisays I looked into this a bit deeper and turns out that there is a route for static content https://github.com/backstage/backstage/blob/master/plugins/techdocs-backend/src/service/router.ts#L143-L145

The really weird bit is that for some reason, Backstage is trying to load techdocs from `https://` when it's not available. This really messes with debugging because without having `https` on localhost it's possible to test it properly. I haven't been able to figure out why this is happening.

**Comment by jennisays** [comment]
It would appear `https` must be hard coded somewhere, and it doesn't look like it's in the router code.

**Comment by Crevil** [comment]
I've hit this as well. It seems to be a browser thing. When querying the URLs with `curl` the redirect location to the static contents are with HTTP.

```
$ curl http://localhost:7000/api/techdocs/docs/default/Component/service-lifecycle/index.html
Found. Redirecting to http://localhost:7000/api/techdocs/static/docs/default/Component/service-lifecycle/index.html%
```

If I follow the redirect, again with curl, it goes through as expected.

```
$ curl -L http://localhost:7000/api/techdocs/docs/default/Component/service-lifecycle/index.html
<!doctype html>
<!-- all the rest -->
```

As noticed in this issue, in the browser (Brave for me) the redirect is upgraded to HTTPs.

This is the initial request that returns `302` with an HTTP `Location` in the response headers. (as expected)

![image](https://user-images.githubusercontent.com/6881694/100011565-31fd5480-2dd2-11eb-8a96-2a1a6c98701b.png)

Brave then issues the redirected request, now on HTTPS.

![image](https://user-images.githubusercontent.com/6881694/100011660-56593100-2dd2-11eb-844b-f54e2d462493.png)

**Comment by Crevil** [comment]
Soo, I think I'm on to something.

`helmet` the webserver we use is configured by default to upgrade insecure connections through the CSP headers.

https://github.com/backstage/backstage/blob/master/packages/backend-common/src/service/lib/ServiceBuilderImpl.ts#L59

By removing this property I'm able to avoid the HTTPS redirect in the browser 🎉 

I can't find a way to remove this default value through configuration however. We seem to merge the custom properties from app-config.yaml into the referenced object.

In other words, I don't know how we should go about fixing it. One solution is off cause just to remove the default value, but that would leave us insecure be default 👎

**Comment by taras** [comment]
@Crevil that's some good sleuthing right there. Very nicely done. 

> One solution is off cause just to remove the default value, but that would leave us insecure be default 👎

This option can't secure the connection if there is no https, it can only upgrade if https connection is available. Would it make it less secure or just less broken?

**Comment by Crevil** [comment]
That's true. But if we omit it, and both options exist (http and https), we won't upgrade.

I guess this os just the trigger really. The root cause sjouæd answer: why does the browser think it can upgrade on the redirect?

**Comment by Crevil** [comment]
I just noticed this also fixes the endless spinners in the bread crumbs 🎉

**Comment by freben** [comment]
Does the following in the `app-config.yaml` fix things?

```yaml
backend:
  csp:
    upgrade-insecure-requests: false
```

You may want to place this in an `app-config.local.yaml` that gets picked up automatically during local dev.

**Comment by taras** [comment]
@freben with the configure you provided, I get the following error `Backend failed to start up, TypeError: Invalid type in config for key 'backend.csp.upgrade-insecure-requests' in 'app-config.local.yaml', got boolean, wanted string-array`

**Comment by freben** [comment]
@taras i actually merged support for that just today :) https://github.com/backstage/backstage/pull/3437

**Comment by freben** [comment]
@Crevil maybe you could try with the most recent backstage and using that configuration option.

**Comment by Crevil** [comment]
Sure thing. This is great 👍 Will see if we can get time in for an upgrade tomorrow and I will test. Thank you 🙌

**Comment by OrkoHunter** [comment]
@Crevil @taras Digging out an old issue from the past :)

As @freben suggests, adding 

```yaml
backend:
  csp:
    upgrade-insecure-requests: false
```

to `app-config.local.yaml` should help you solve the automatic HTTP -> HTTPS overwrite. 

For the https://demo.backstage.io site hosted at https://github.com/backstage/demo - we use a similar setup where we deploy the frontend bundle on backend - not noticing any problems.

TBH I'm not sure what to do with this issue.

**Comment by Crevil** [comment]
Sorry for never comming back to this issue. The discussed solutions did work and following upgrades have worked as well. I think we can close this off if @taras  has not anything else on it?

**Comment by taras** [comment]
@Crevil yep, agree.
